waiter/src/waiter/statsd.clj

```diff
@@ -119,255 +174,334 @@
+          (percentiles values 50 75 95 99 100)))
+
+  (defn bounded-conj
+    "Like conj, except it will start to pop after the count reaches histogram-max-size.
+     We use this with PersistentQueues, which conj onto the rear and pop from the front,
+     to get sliding queue behavior for histograms."
+    [queue value metric-group]
+    (let [queue' (conj queue value)
+          limit (or (:histogram-max-size @config) default-histogram-max-size)]
+      (if (> (count queue') limit)
+        (do
+          (counters/inc! (metrics/waiter-counter "statsd" "histo-values-dropped" metric-group))
+          (pop queue'))
+        queue')))
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; helpers for getting datadog api tag arrays
+
+  (defn- dd-tags
+    "Return a string array of datadog tags,
+     for passing to the varargs String... argument
+     as expected by all methods on the java dogstatsd client."
+    ^"[Ljava.lang.String;" ; return-type hint as String[]
+    [config tags-key metric-group]
+    (let [metric-group-tag (str "metric_group:" metric-group)
+          base-tags (get config tags-key)
+          tags (conj base-tags metric-group-tag)]
+      (into-array String tags)))
+
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+;; metric type config definitions
+
+  (def set-metric
+    "Before sending to Statsd, we need to convert to an integer,
+     so we take the hashCode here. We aggregate sets by union-ing."
+    {:seed-fn (fn [^Object value] #{(.hashCode value)})
+     :dogstatsd-client-fn (fn dd-set [config client metric-group metric value]
+                            (.recordSetValue ^StatsDClient client ^String metric (str value)
+                              (dd-tags config :server-tags metric-group)))
+     :aggregate-fn (fn [values-set ^Object value _] (conj values-set (.hashCode value)))
+     :publish-fn set-publish
+     :delete-on-publish?-fn (constantly true)
+     :metric-type :set})
+
+  (def gauge-delta-metric
+    "Gauge deltas aggregate by addition (like counters), but we
+     only want to delete on publish when the value is 0."
+    {:seed-fn identity
+     :aggregate-fn (fn [sum value _] (+ sum value))
+     :dogstatsd-client-fn (fn dd-gauge-delta [config client metric-group metric value]
+                            (.count ^StatsDClient client ^String metric (long value)
+                              (dd-tags config :server-tags metric-group)))
+     :publish-fn gauge-delta-publish
+     :delete-on-publish?-fn zero?
+     :metric-type :gauge-delta})
+
+  (def gauge-metric
+    "Gauges aggregate by replacing the old value with the new one"
+    {:seed-fn identity
+     :aggregate-fn (fn [_ value _] value)
+     :dogstatsd-client-fn (fn dd-gauge [config client metric-group metric value]
+                            (.recordGaugeValue ^StatsDClient client ^String metric (double value)
+                              (dd-tags config :cluster-tags metric-group)))
```

Comment:
@shamsimam ^^^ this is the only place we use `:cluster-tags`, because this is also the only metric-type that was using `metric-path-without-server` to build the long metric name.